akka-http/src/main/java/akka/http/javadsl/server/HttpApp.java

```diff
@@ -0,0 +1,150 @@
+        .toCompletableFuture()
+        .exceptionally(ignored -> BoxedUnit.UNIT) // If the future fails, we want to complete normally
+        .get(); // It's waiting forever because maybe there is never a shutdown signal
+    } finally {
+      bindingFuture.thenCompose(ServerBinding::unbind).handle( (success, exception) -> {
+        postServerShutdown(exception, theSystem);
+        if (!system.isPresent()) {
+          // we created the system. we should cleanup!
+          theSystem.terminate();
+        }
+        return null;
+      });
+    }
+
+  }
+
+  /**
+   * It tries to retrieve the {@link ServerBinding} if the server has been successfully started. It throws an {@link IllegalStateException} otherwise.
+   * You can use this method to attemp to retrieve the {@link ServerBinding} at any point in time to, for example, stop the server due to unexpected circumstances.
+   */
+  ServerBinding binding() {
+    if (serverBinding.get() == null)
+      throw new IllegalStateException("Binding not yet stored. Have you called startServer?");
+    return serverBinding.get();
+  }
+
+  /**
+   * Hook that will be called just after the server termination. Override this method if you want to perform some cleanup actions after the server is stopped
+   * When problems shutting down {@code possibleFailure} contains the failure occurred. It is null otherwise
+   */
+  protected void postServerShutdown(Throwable possibleFailure, ActorSystem system) {
+    System.out.println("Shutting down the server");
+  }
+
+  /**
+   * Hook that will be called just after the Http server binding is done. Override this method if you want to perform some actions after the server is up
+   */
+  protected void postHttpBinding (ServerBinding binding) {
+    System.out.println("Server online at http://" + binding.localAddress().getHostName() + ":" + binding.localAddress().getPort() + "/");
+  }
+
+  /**
+   * Hook that will be called in case the Http server binding fails. Override this methid if you want to perform some actions after the server binding failed
+   */
+  protected void postHttpBindingFailure(Throwable cause) {
+    System.out.println("Error starting the server: " + cause.getMessage());
+    cause.printStackTrace(System.err);
+  }
+
+  /**
+   * Hook that lets the user specify the future that will signal the shutdown of the server whenever completed
+   */
+  protected CompletionStage<BoxedUnit> waitForShutdownSignal (ActorSystem system) {
+    return CompletableFuture.supplyAsync( () -> {
+      System.out.println("Press RETURN to stop...");
+      try {
+        System.in.read();
+      } catch (IOException e) {
+        System.out.println("Problem occurred! " + e.getMessage());
+        e.printStackTrace(System.err);
+      }
+      return BoxedUnit.UNIT;
+    });
+  }
+
+  /**
+   * Your class should provide an implementation for the routes. These will be bound to the [[Http]]
```

Comment:
Try to avoid "you" in scaladoc. Say "Override to implement route that will be served by this http server.